src/veriff/veriff.service.ts

```diff
@@ -0,0 +1,175 @@
+  constructor(
+    private readonly credentialFactory: VerifiableCredentialFactory,
+    private readonly veriffWatchlistService: VeriffWatchlistService,
+    private readonly logger: Logger,
+    private readonly communicationGateway: CommunicationGateway,
+  ) {}
+  private readonly VERIFF_SECRET =
+    process.env.VERIFF_SECRET || 'default_secret'; // .env usage
+
+  validateSignature(signature: string, payload: any): boolean {
+    const computedSignature = crypto
+      .createHmac('sha256', this.VERIFF_SECRET)
+      .update(JSON.stringify(payload))
+      .digest('hex');
+
+    return computedSignature === signature;
+  }
+
+  async validateWebhookRequest(
+    signature: string,
+    payload: VeriffWebhookPayload,
+  ): Promise<void> {
+    this.logger.debug('Handling webhook payload from Veriff:', payload);
+    const { vendorData: jwt, data, status } = payload;
+    if (!jwt) {
+      this.logger.warn('vendorData (VC JWT) is missing, cannot proceed.');
+      throw new BadRequestException(
+        'vendorData (VC JWT) is missing, cannot proceed.',
+      );
+    }
+
+    // Verify signature
+    if (!this.validateSignature(signature, payload)) {
+      this.logger.warn('Invalid signature, cannot proceed.');
+      throw new UnauthorizedException('Invalid Veriff signature.');
+    }
+
+    try {
+      const vc = await this.credentialFactory.create<VeriffPayload>(jwt);
+
+      const did = vc.getId();
+
+      if (!did) {
+        this.logger.warn('VC is missing DID, cannot proceed.');
+        throw new BadRequestException('Verifiable Credential is missing DID.');
+      }
+
+      // Check pepSanctionMatches
+      if (data.verification.decision === 'approved') {
+        try {
+          let pepSanctionMatches: WatchlistScreeningResult | undefined;
+          if (ENABLE_PEP_CHECK) {
+            pepSanctionMatches =
+              await this.veriffWatchlistService.getWatchlistScreening(
+                payload.sessionId,
+              );
+            this.logger.debug('Watchlist result:', pepSanctionMatches);
+          }
+        } catch (e) {
+          this.logger.warn('Failed to fetch watchlist screening:', e.message);
+        }
+
+        const issuer = await getTonomyOpsIssuer();
+
+        const person = data.verification.person;
+
+        const firstName = getFieldValue(person, 'firstName');
+        const lastName = getFieldValue(person, 'lastName');
+        const birthDate = getFieldValue(person, 'dateOfBirth');
+        const nationality = getFieldValue(person, 'nationality');
+
+        if (!firstName || !birthDate || !nationality) {
+          throw new Error('Missing required personal data.');
+        }
+
+        const signedFirstNameVc = await util.VerifiableCredential.sign(
+          '',
+          'FirstNameCredential',
+          { firstName },
+          issuer,
+          { subject: did },
+        );
+
+        const signedLastNameVc = await util.VerifiableCredential.sign(
+          '',
```

Comment:
better to make an id instead of ''.
use `https://tonomy.foundation/vc/id/{random 8 character string}` like this https://github.com/Tonomy-Foundation/Tonomy-ID-SDK/blob/master/src/sdk/util/ssi/vc.ts#L285